src/main/kotlin/cacophony/grammar/syntaxtree/Expression.kt

```diff
@@ -0,0 +1,253 @@
+package cacophony.grammar.syntaxtree
+
+import cacophony.utils.Location
+
+sealed class Expression(
+    val range: Pair<Location, Location>,
+) { // everything in cacophony is an expression
+    // artificial instance, can be useful when calculating values of nested expressions
+    class Empty(
+        range: Pair<Location, Location>,
+    ) : Expression(range)
+
+    class VariableUse(
+        range: Pair<Location, Location>,
+        val identifier: String,
+    ) : Expression(range)
+
+    sealed class Type(
+        range: Pair<Location, Location>,
+    ) {
+        class Basic(
+            range: Pair<Location, Location>,
+            val identifier: String,
+        ) : Type(range)
+
+        class Functional(
+            range: Pair<Location, Location>,
+            vararg val argumentsType: Type,
```

Comment:
This will require a ton of asList() and spread operator calls in every function dealing with AST. Can't we just use a normal List?

Reply:
I up that, we have similar interface in AlgebraicRegex and it was pain to cast everything to TypedArray every time. The vararg constructor is nice for testing, so I propose we add it as non-primary one.